deepmd/pt/utils/tabulate.py

```diff
@@ -0,0 +1,971 @@
+                    yy, torch.from_numpy(self.matrix["layer_" + str(layer + 1)][idx])
+                ) + torch.from_numpy(self.bias["layer_" + str(layer + 1)][idx])
+                if self.neuron[layer] == self.neuron[layer - 1]:
+                    zz = (
+                        self._layer_0(
+                            yy,
+                            self.matrix["layer_" + str(layer + 1)][idx],
+                            self.bias["layer_" + str(layer + 1)][idx],
+                        )
+                        + yy
+                    )
+                    dz = unaggregated_dy_dx(
+                        zz - yy,
+                        self.matrix["layer_" + str(layer + 1)][idx],
+                        dy,
+                        ybar,
+                        self.functype,
+                    )
+                    dy2 = unaggregated_dy2_dx(
+                        zz - yy,
+                        self.matrix["layer_" + str(layer + 1)][idx],
+                        dy,
+                        dy2,
+                        ybar,
+                        self.functype,
+                    )
+                elif self.neuron[layer] == 2 * self.neuron[layer - 1]:
+                    tt, zz = self._layer_1(
+                        yy,
+                        self.matrix["layer_" + str(layer + 1)][idx],
+                        self.bias["layer_" + str(layer + 1)][idx],
+                    )
+                    dz = unaggregated_dy_dx(
+                        zz - tt,
+                        self.matrix["layer_" + str(layer + 1)][idx],
+                        dy,
+                        ybar,
+                        self.functype,
+                    )
+                    dy2 = unaggregated_dy2_dx(
+                        zz - tt,
+                        self.matrix["layer_" + str(layer + 1)][idx],
+                        dy,
+                        dy2,
+                        ybar,
+                        self.functype,
+                    )
+                else:
+                    zz = self._layer_0(
+                        yy,
+                        self.matrix["layer_" + str(layer + 1)][idx],
+                        self.bias["layer_" + str(layer + 1)][idx],
+                    )
+                    dz = unaggregated_dy_dx(
+                        zz,
+                        self.matrix["layer_" + str(layer + 1)][idx],
+                        dy,
+                        ybar,
+                        self.functype,
+                    )
+                    dy2 = unaggregated_dy2_dx(
+                        zz,
+                        self.matrix["layer_" + str(layer + 1)][idx],
+                        dy,
+                        dy2,
+                        ybar,
+                        self.functype,
+                    )
+                dy = dz
+                yy = zz
+
+        vv = zz.detach().numpy()
+        dd = dy.detach().numpy()
+        d2 = dy2.detach().numpy()
```

Comment:
You can use this method. I am wondering whether `cpu()` is critical...

https://github.com/deepmodeling/deepmd-kit/blob/0fa1b435355cbcfe10af8e766a38fe34c502738f/deepmd/pt/utils/utils.py#L56-L73

Reply:
I think it's OK, all the computations are done on the CPU